x-pack/plugin/mapper-counted-keyword/src/main/java/org/elasticsearch/xpack/countedkeyword/CountedTermsAggregator.java

```diff
@@ -0,0 +1,203 @@
+import java.util.Arrays;
+import java.util.Map;
+import java.util.function.BiConsumer;
+import java.util.function.Supplier;
+
+import static java.util.Collections.emptyList;
+import static org.apache.lucene.index.SortedSetDocValues.NO_MORE_ORDS;
+import static org.elasticsearch.search.aggregations.InternalOrder.isKeyOrder;
+
+class CountedTermsAggregator extends TermsAggregator {
+    private final BytesKeyedBucketOrds bucketOrds;
+    protected final ValuesSource.Bytes.WithOrdinals valuesSource;
+
+    @SuppressWarnings("this-escape")
+    CountedTermsAggregator(
+        String name,
+        AggregatorFactories factories,
+        ValuesSource.Bytes.WithOrdinals valuesSource,
+        BucketOrder order,
+        DocValueFormat format,
+        BucketCountThresholds bucketCountThresholds,
+        AggregationContext context,
+        Aggregator parent,
+        CardinalityUpperBound cardinality,
+        Map<String, Object> metadata
+    ) throws IOException {
+        super(name, factories, context, parent, bucketCountThresholds, order, format, SubAggCollectionMode.DEPTH_FIRST, metadata);
+        this.valuesSource = valuesSource;
+        this.bucketOrds = BytesKeyedBucketOrds.build(context.bigArrays(), cardinality);
+    }
+
+    @Override
+    public LeafBucketCollector getLeafCollector(AggregationExecutionContext aggCtx, LeafBucketCollector sub) throws IOException {
+        SortedSetDocValues ords = valuesSource.ordinalsValues(aggCtx.getLeafReaderContext());
+        return new LeafBucketCollectorBase(sub, ords) {
+
+            @Override
+            public void collect(int doc, long owningBucketOrd) throws IOException {
+                if (ords.advanceExact(doc) == false) {
+                    return;
+                }
+                for (long ord = ords.nextOrd(); ord != NO_MORE_ORDS; ord = ords.nextOrd()) {
+                    long bucketOrdinal = bucketOrds.add(owningBucketOrd, ords.lookupOrd(ord));
+                    if (bucketOrdinal < 0) { // already seen
+                        bucketOrdinal = -1 - bucketOrdinal;
+                        collectExistingBucket(sub, doc, bucketOrdinal);
+                    } else {
+                        collectBucket(sub, doc, bucketOrdinal);
+                    }
+                }
+            }
+        };
+    }
+
+    @Override
+    public InternalAggregation[] buildAggregations(long[] owningBucketOrds) throws IOException {
+        StringTerms.Bucket[][] topBucketsPerOrd = new StringTerms.Bucket[owningBucketOrds.length][];
+        long[] otherDocCounts = new long[owningBucketOrds.length];
+        for (int ordIdx = 0; ordIdx < owningBucketOrds.length; ordIdx++) {
+            collectZeroDocEntriesIfNeeded(owningBucketOrds[ordIdx]);
+            int size = (int) Math.min(bucketOrds.size(), bucketCountThresholds.getShardSize());
+
+            BucketPriorityQueue<StringTerms.Bucket> ordered = new BucketPriorityQueue<>(size, partiallyBuiltBucketComparator);
```

Comment:
We don't need to support an explicit order for this use case and buckets descendingly ordered by doc count sounds fine.

Reply:
I see then this code makes sense. It would allow other orders as well, but given that the builder doesn't expose the order option this would never happen. Maybe add a comment that in practice we always sort by doc count descending?